source/class/qx/tool/migration/BaseMigration.js

```diff
@@ -0,0 +1,225 @@
+
+    /**
+     * @see {@link qx.tool.config.Utils#getQxVersion}
+     */
+    getQxVersion: qx.tool.config.Utils.getQxVersion.bind(qx.tool.config.Utils),
+
+    /**
+     * @see {@link qx.tool.config.Utils#getQxAppVersion}
+     */
+    getAppQxPath: qx.tool.config.Utils.getAppQxPath.bind(qx.tool.config.Utils),
+
+    /**
+     * Returns the version of qooxdoo this migration applies to.
+     */
+    getVersion() {
+      return this.classname.match(/\.M([0-9_]+)$/)[1].replace(/_/g,".");
+    },
+
+    /**
+     * Output message that announces a migration. What this does is to mark it
+     * visually
+     * @param message
+     */
+    announce(message) {
+      if (this.getRunner().getVerbose()) {
+        this.info("*** " + message);
+      }
+    },
+
+    /**
+     * Rename source files
+     * @param {String[]} fileList Array containing arrays of [new name, old name]
+     * @return {Promise<Boolean>} Whether this migration still has to be applied
+     */
+    async renameFiles(fileList) {
+      let migrationInfo = this.getRunner().createMigrationInfo();
+      let dryRun = this.getRunner().getDryRun();
+      qx.core.Assert.assertArray(fileList);
+      let filesToRename = this.checkFilesToRename(fileList);
+      if (filesToRename.length) {
+        if (dryRun) {
+          // announce migration
+          this.announce(`The following files will be renamed:`);
+          for (let [newPath, oldPath] of filesToRename) {
+            this.announce(`'${oldPath}' => '${newPath}'.`);
+          }
+          migrationInfo.pending++;
+        } else {
+          // apply migration
+          for (let [newPath, oldPath] of filesToRename) {
+            try {
+              await fs.renameAsync(oldPath, newPath);
+              this.debug(`Renamed '${oldPath}' to '${newPath}'.`);
+            } catch (e) {
+              this.error(`Renaming '${oldPath}' to '${newPath}' failed: ${e.message}.`);
+              process.exit(1);
+            }
+          }
+          migrationInfo.applied++;
+        }
+      }
+      return migrationInfo;
+    },
+
+    /**
+     * Given an array of [newPath,oldPath], filter by those which exist
+     * at oldPath and not at newPath
+     * @param fileList {[]}
+     * @return []
+     */
+    checkFilesToRename(fileList) {
+      let filesToRename = [];
+      for (let [newPath, oldPath] of fileList) {
+        if (!fs.existsSync(newPath) && fs.existsSync(oldPath)) {
+          filesToRename.push([newPath, oldPath]);
+        }
+      }
+      return filesToRename;
+    },
+
+    /**
+     * Checks if the given file or array of files contains a given text
+     * @param {String|String[]} files
+     * @param {String} text
+     * @return {Promise<Boolean>}
+     */
+    async checkFilesContain(files, text) {
+      files = Array.isArray(files) ? files : [files];
+      for (let file of files) {
+        if ((await fsp.stat(file)).isFile() && (await fsp.readFile(file, "utf8")).includes(text)) {
```

Comment:
This looks to me like if `fsp.stat().isFile()` returns false, we'll be doing `if false.includes(text)` which will throw. Is that what's intended?

Reply:
As per the Docs, the method always returns a <Stats> object. https://nodejs.org/api/fs.html#fs_class_fs_stats even if there is no file.

Reply:
That's ugly. A function that begins "is" generally returns a boolean. Anyway, not our problem.